keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -34,235 +34,172 @@ status: provisional
+- the taint Value is string now, it is theoretically possible that several errors are detected for one PV, we may extend the string to cover this situation: combine the errors together and splited by semicolon or other symbols.
 
-For unused volumes, the volume manager will try to mount again when we want to use it, so we do not need any effect.
-
-In summary, we set the VolumeTaintEffect to NoEffect too.
-
-**The last one is: `VolumeNotUsable` taint**,
-```
-Key: VolumeNotUsable (e.g. bad blocks, filesystem corrupted, etc)
-Value: true
-VolumeTaintEffect: NoAttachOrMount
-```
 
-If the volume is not usable because of bad blocks, filesystem corruption, disk deletion or any other reasons, we need to taint the PV using this taint.
+### CSI change
 
-Here, we use NoAttachOrMount effect mainly for multi-attach volumes. 
+Container Storage Interface (CSI) specification will be modified to add two RPCs ControllerCheckVolume and NodeCheckVolume. 
 
-Because NoAttachOrMount effect will prevent new pods from using the tainted PV. 
-
-And as mentioned above, for in use volumes, reactions are also needed.
-
-When we finish discussing the reaction for unhealthy PVs, we need to revisit all the VolumeTaintEffects.
-
-### Networked storage
-
-For networked storage, the following volume health conditions will be checked:
-* Checks if volume still exists and/or is attached
-* Checks if volume is still mounted and usable
-
-Container Storage Interface (CSI) specification will be modified to add two RPCs ControllerCheckVolume and NodeCheckVolume. A monitor controller (external-monitor) that is responsible for watching the PersistentVolumeClaim, PersistentVolume, and VolumeAttachment API objects will be added. The external-monitor can be deployed as a sidecar container together with the CSI driver.  For PVs, the external-monitor calls CSI driver’s monitor interface to check volume health status.
-
-* ControllerCheckVolume RPC
-  * It calls ControllerCheckVolume() to see if volume still exists; Taints PV with VolumeNotExisting if not existing any more.
-  * If volume is attached, it calls ControllerCheckVolume() to see if volume is still attached; Taints PV with VolumeNotAttached if not attached any more.
-
-* NodeCheckVolume RPC
-  * For any PVC that is mounted, the external-monitor calls NodeCheckVolume() to see if volume is still mounted; Taints PV with VolumeNotMounted if not mounted any more.
-  * Calls NodeCheckVolume to check if volume is usable, e.g., filesystem corruption, bad blocks, etc; Taint PV with VolumeNotUsable if volume is not usable.
+- ControllerCheckVolume RPC
+  - External controller calls ControllerCheckVolume() to check the health condition of PVs themselves, for example, if the PVs are deleted ...
+  - Move Attach checking here ?
+- NodeCheckVolume RPC
+  - If volume is attached to this node, the external agent calls NodeCheckVolume() to see if volume is still attached; Move this check to  ControllerCheckVolume ?
```

Comment:
We can keep this here.  In the case of multi-attach, it is easier to check on each node as discussed earlier.